api/src/auth/drivers/saml.ts

```
@@ -0,0 +1,178 @@
+			.first();
+
+		return user?.id;
+	}
+
+	async getUserID(payload: Record<string, any>) {
+		const { provider, emailKey, identifierKey, nameKey, givenNameKey, familyNameKey, allowPublicRegistration } =
+			this.config;
+
+		const email = payload[emailKey ?? 'http://schemas.xmlsoap.org/ws/2005/05/identity/claims/emailaddress'];
+		const identifier = payload[identifierKey ?? 'http://schemas.xmlsoap.org/ws/2005/05/identity/claims/nameidentifier'];
+		const userID = await this.fetchUserID(identifier);
+
+		const userInfo = {
+			name: payload[nameKey ?? 'http://schemas.xmlsoap.org/ws/2005/05/identity/claims/name'],
+			given_name: payload[givenNameKey ?? 'http://schemas.xmlsoap.org/ws/2005/05/identity/claims/givenname'],
+			family_name: payload[familyNameKey ?? 'http://schemas.xmlsoap.org/ws/2005/05/identity/claims/surname'],
+			email,
+		};
+
+		if (userID) return userID;
+
+		// Is public registration allowed?
+		if (!allowPublicRegistration) {
+			logger.trace(`[SAML] User doesn't exist, and public registration not allowed for provider "${provider}"`);
+			throw new InvalidCredentialsException();
+		}
+
+		try {
+			await this.usersService.createOne({
+				provider,
+				first_name: userInfo.given_name,
+				last_name: userInfo.family_name,
+				email: email,
+				external_identifier: identifier.toLowerCase(),
+				role: this.config.defaultRoleId,
+			});
+			const userID = await this.fetchUserID(identifier);
+			if (userID) return userID;
+		} catch (e) {
+			if (e instanceof RecordNotUniqueException) {
+				logger.warn(e, '[SAML] Failed to register user. User not unique');
+				throw new InvalidProviderException();
+			}
+			throw e;
+		}
+
+		throw new InvalidCredentialsException();
+	}
+
+	async login(user: User): Promise<void> {
+		let authData = user.auth_data as AuthData;
+
+		if (typeof authData === 'string') {
+			try {
+				authData = parseJSON(authData);
+			} catch {
+				logger.warn(`[SAML] Session data isn't valid JSON: ${authData}`);
+			}
+		}
+	}
```

I don't quite understand this function. It seems we never set `auth_data`. Is this ever used, and if so, wouldn't it be better to throw is the data was invalid?

---

We don't use it you can correct, can change to just a throw instead.

---

But when do we set `auth_data`? This should never trigger, right?